core/src/main/scala/kafka/network/SocketServer.scala

```diff
@@ -548,21 +562,23 @@ private[kafka] class Processor(val id: Int,
         requestChannel.sendRequest(req)
         selector.mute(receive.source)
       } catch {
-        case e @ (_: InvalidRequestException | _: SchemaException) =>
-          // note that even though we got an exception, we can assume that receive.source is valid. Issues with constructing a valid receive object were handled earlier
-          error(s"Closing socket for ${receive.source} because of error", e)
-          close(selector, receive.source)
+        // note that even though we got an exception, we can assume that receive.source is valid. Issues with constructing a valid receive object were handled earlier
+        case e: Throwable => processChannelException(receive.source, s"Exception while processing request from ${receive.source}", e)
       }
     }
   }
 
   private def processCompletedSends() {
     selector.completedSends.asScala.foreach { send =>
-      val resp = inflightResponses.remove(send.destination).getOrElse {
-        throw new IllegalStateException(s"Send for ${send.destination} completed, but not in `inflightResponses`")
+      try {
+        val resp = inflightResponses.remove(send.destination).getOrElse {
+          throw new IllegalStateException(s"Send for ${send.destination} completed, but not in `inflightResponses`")
+        }
+        updateRequestMetrics(resp.request)
+        selector.unmute(send.destination)
+      } catch {
+        case e: Throwable => processChannelException(send.destination, s"Exception while processing completed send to ${send.destination}", e)
```

Comment:
Do we have these catch blocks in case of bugs in Selector or is it to catch things like OutOfMemoryErrors? Either way, there are 2 things we need to think about:

1. We do `connectionQuotas.dec` during `close`, but we don't remove the connection id from `inflightResponses`. That seems a bit inconsistent.

2. We have to be careful not to double count disconnections when decrementing `connectionQuotas`. This could potentially be an issue if a channel appears in `disconnected` and a Selector method for that channel also throws an exception. Because we only close the channel in `processChannelException` if the channel is still in the `Selector`, this may be OK, but it would be good to double-check, document and test.

Reply:
The catch block matches everything we used to catch before in `Processor` to avoid terminating `Processor` and attempts cleanup of channels in each case. This happens to include IllegalStateException (bug) as well as OutOfMemoryError. Only ControlThrowables are propagated.
1. Added `inflightResponses.remove`
2. Have added unit tests in `Selector` and `SocketServer` to ensure disconnect notifications are not generated when the channel is explicitly closed. And as you said, we do check to avoid double counting.

Reply:
I one change I made to the actual flow is the handling of `IOException` and `IllegalStateException` from `Selector.poll()`. I felt the code was odd because it closed `Selector`, but left the processor thread running, polling etc. I changed it to log the error and carry on, which isn't ideal either, because `poll()` may fail again, but it fits in with the comment about not terminating `Processor`.

Reply:
Nit: long line.